include/util/workqueue.h

```diff
@@ -0,0 +1,211 @@
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    1. Redistributions of source code must retain the above copyright notice, this
+       list of conditions and the following disclaimer.
+    2. Redistributions in binary form must reproduce the above copyright notice,
+       this list of conditions and the following disclaimer in the documentation
+       and/or other materials provided with the distribution.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#pragma once
+
+#include <thread>
+
+RENGINE_BEGIN_NAMESPACE
+
+/*!
+    The WorkQueue class encapsulates a queue of jobs that will be run on a
+    background thread.
+
+    Jobs are schedule for execution by calling WorkQueue::schedule().
+
+    Jobs are implemented by reimplementing WorkQueue::Job and the Job::onRun()
+    virtual function. Jobs will be completed in their own time, but the job
+    can be checked for completion on the calling thread through
+    Job::isCompleted() and it is possible to wait for a job to complete with
+    Job::waitForCompletion().
+ */
+
+class WorkQueue
+{
+public:
+
+    class Job
+    {
+    public:
+        virtual ~Job() { }
+
+        /*!
+            Returns if the job has completed or not. This function is thread-
+            safe.
+         */
+        bool isCompleted() const;
+
+        /*!
+            Blocks until the job has completed. The function returns
+            immediately, if the job has already been completed. This function
+            is thread-safe.
+         */
+        void waitForCompletion();
+
+        /*!
+            Custom jobs should override this function to perform jobs on the
+            work queue. The job will be marked as completed immediately after
+            the onRun() function completes.
+         */
+        virtual void onRun() { }
+
+    private:
+        friend class WorkQueue;
+
+        /*!
+            This function is called by the WorkQueue to execute the job,
+            meaning, calling it's onRun() function, setting the completed
+            state and signalling threads waiting for its completion.
+         */
+        void execute();
```

Comment:
Would it make sense to do this inline in run()? Then it doesn't "clutter" the API since it's a bit small - I initially got a bit confused why I wouldn't be able to start this since it was private, but then realised it's just a helper of sorts